### DISCUSSION TITLE
Add cost tracking to utility meters

### DISCUSSION THREAD

**Discussion by langestefan** [home-assistant/architecture#1081] [Entity Models]
Tracking costs for individual devices makes a lot of sense. 

For those of us on multiple tariffs or dynamic hourly energy contracts energy costs cannot directly be calculated from energy consumption, it requires knowing when the energy was consumed.

There is currently no good way to do so without cooking up something yourself using YAML. 

Adding this functionality to the utility meter seems like the perfect place to do it since it already has all the required information. Adding cost tracking to utility meters can also serve as a stepping stone to enable cost tracking for individual devices in the energy dashboard, see https://github.com/home-assistant/architecture/discussions/790. I think the solution proposed there of defining the concept of multiple grids is not feasible, since the complexity of supporting all possible configurations will explode quickly. 

To build a cost tracking function for utility meters I propose the following:

- For each individual tariff in the utility meter we add one optional cost field.
- Each tariff will have an associated input for fixed cost / unit consumed, or an entity supplying the current cost / unit consumed. Essentially the same idea as the energy dashboard cost tracking.
- The cost entity should be named something like `sensor.<utility-meter-name>_<tariff-name>_cost`
- If the cost input for a particular tariff is not supplied the corresponding cost entity will not be created.
- If the utility meter has no tariffs it should still be possible to set a cost / unit, which then applies to everything the sensor records.
- There should be a total cost entity, which is simply the running sum of all the individual costs combined.
- Cost entities will not be limited to energy device classes since defining a generic 'cost / unit' will also enable other usecases. For example tracking costs of something that runs for x hours with a known cost / hour of runtime.

**Comment by AJediIAm** [comment]
I would very much like to have the option of tracking cost of devices based on dynamic (hourly) energy prices. 

Adding a single (optional) attribute to the utility meter for the current cost would be enough so the utility meter can track an attribute with the cost in addition to its original functionality.

I don't think the logic of multiple tariffs should be in the utility meter. There are too many complex situations which are impossible to capture. Having a template with the current cost seems a very reasonable compromis to keep the utility meter simple.

**Comment by AJediIAm** [comment]
I'm not an expert, but I think your use cases are already supported by the utility meter: https://www.home-assistant.io/integrations/utility_meter/

You can create a utility meter to track your energy usage and use the service UTILITY_METER.CALIBRATE to set a negative value reflecting the prepaid amount. The utility meter will show how much energy you have left.

It's one of the more difficult helpers to fully understand. The documentation already contains some good examples. Maybe these van be extended with a few more scenarios like the prepaid energy use case.

**Reply by langestefan** [reply]
Huh? What does this have to do with cost tracking?

**Reply by dgomes** [reply]
utility meter can be used to track *anything* it can be kWh or $/€

**Reply by langestefan** [reply]
this is not for tracking $/€ but for _creating_ a cost statistic, so it's a completely  different usecase as to what you describe. But perhaps it's out of scope for the utility meter, but then we need to create something else.

**Reply by AJediIAm** [reply]
Let's remember that we're all on the same team here.

@langestefan has a challenge/problem.
Either there is a fitting solution already and we can update the documentation on how to use it for the described use case, or we come to a precise description and business case for improving home assistant. Win-win.

Reading back, I see several use cases being discussed and I think we have gone off track from the original use case. Let's assume some use cases can be solved with the utility meter and some use cases are still unsolved.
It requires going back and forward, experimenting and explaining to try out different ideas until we have a clear picture of what @langestefan needs exactly and where the current utility meter falls short.

Where do we stand:
1) I proposed to use a single entity for the current tariff. Countries have different tariffs and it's too complicated to capture this complexity in the utility meter.
The template sensor can be used to create an entity with the current energy price.
Is this acceptable?

2) The current utility meter can be used to track cost per device using the tariff entity. Home Assistant automatically keeps track of statistics.
I understand this is not exactly what you need, but I don't understand yet which information is still missing or how does this fall short? Can you provide a comparison of the current utility meter and a mock-up of what you want to see?

**Reply by langestefan** [reply]
Did ChatGPT write this? 'Where do we stand'? Are you joking?

**Reply by AJediIAm** [reply]
Be nice. I'm trying to help you.

**Reply by langestefan** [reply]
You're not helping

**Reply by EDelsman** [reply]
I think there's a misunderstanding. The question is "how to sum up the cost of the energy that is tracked by the utility meter". 

That is not the same as applying a utility meter to a total cost sensor. The problem at hand is: how to get a total cost sensor in the first place. 

That is not trivial because the price may vary over time. The logic to do so is already present in the energy dashboard, but not available to separate entities. The suggestion is to have the utility meter calculate that cost. 

The tariff in the utility meter is not suited for this goal, because it just splits the energy in parts per tariff.

Personally I think assigning the task of calculating the cost should not be in the utility meter, but in a new, separate type of entity. Then, as suggested, the utility meter can determine cost per period from this other sensor, without any alteration to the current utility meter functionality.

Putting the cost calculation inside the utility meter would pollute the original goal of the entity.
